### ISSUE TITLE
The WASM specification is not human-readable

### ISSUE THREAD

**Issue by nta** [WebAssembly/spec#983]
Hi,

I've been trying to understand the WebAssembly specification, however was completely baffled by the use of mathematical notation and all the 'weird symbols' to define behavior of the specification.

Take for example [this section](https://webassembly.github.io/spec/core/bikeshed/index.html#integer-operations%E2%91%A0) (and it's not even one of the worst):

![image](https://dotbas.net/i/08d6/holomastigote.png)

To me, as a human being with experience in computer programming, but who hasn't had the luck of having had a formal education in mathematics and related.. things, I feel excluded by the fact that practically every sentence in the specification is human-readable, except for the actual definitions of 'what is supposed to happen'.

I don't think it is a valid use case to write a specification for the presumably <10% of the world population that knows high-level mathematics, especially seeing that it is perfectly possible to have an interest (and knowledge) in software development without being a 'math genius'.

Other specifications I've seen write things like this in human-readable pseudocode using plain English characters, and not having to resort to this unreadable jumble of math symbols and two-dimensional notation.

**Comment by bitwalker** [comment]
I'm not trying to suggest that you may not have a point, but what part of your example is high-level mathematics? The paragraph before defines `N` (i.e. the bit width of the integer), and the rest is a simple formula for calculating the value, with two possibilities depending on which predicate is satisfied (i.e. the part in parentheses), and even explains that it is a two's complement representation for additional clarity.

I agree if there are unfamiliar operators or symbols in the spec, that they should absolutely be defined, but I'm not sure it is valid criticism to say that symbols/operators which see extremely broad use are not "human readable" (e.g. the symbols/operators in your example).

Perhaps there is a better example to make the argument with?

**Comment by ericprud** [comment]
I suspect that this could be address by a couple examples of how to read the spec in the bottom of [§ 2.1. Conventions](https://webassembly.github.io/spec/core/bikeshed/index.html#conventions), after the conventions are introduced.

BTW, the "[this section](https://webassembly.github.io/spec/core/bikeshed/index.html#integer-operations)" link above has a spurious '①' (%E2%91%A0) at the end of it 'cause the '§' links stick it on there. I'll raise a separate issue.

**Comment by nta** [comment]
> Perhaps there is a better example to make the argument with?

I'm not sure if this section is meant to be understood by anyone not trying to 'formally validate' (whatever that might mean) the specification, but [here's supposed to be another example](https://www.w3.org/TR/wasm-core-1/#host-function-instances--hrefsyntax-funcinstmathsftypehrefsyntax-functypemathitfunctype-hrefsyntax-funcinstmathsfhostcodemathithf).

![what?!](https://dotbas.net/i/6bea/colpohyperplasia.png)

**Comment by gahaas** [comment]
The spec is written both in prose as well as with the mathematical notations. There is no need to read the mathematical notations, just stick to the prose text. When you implement the spec, then the prose text is sometimes unclear. In that case, the mathematical notations bring clarity.

**Comment by bitwalker** [comment]
@nta That notation is explained in plain english right above:

- The function type functype must be valid.
- Let [t1∗​]→[t2∗​] be the function type functype.
- For every valid store S1​ extending S and every sequence val∗ of values whose types coincide with t1∗​:
  - Executing hf in store S1​ with arguments val∗ has a non-empty set of possible outcomes.
  - For every element R of this set:
    - Either R must be ⊥ (i.e., divergence).
    - Or R consists of a valid store S2​ extending S1​ and a result result whose type coincides with [t2∗​]. 
- Then the function instance is valid with function type functype.

**Comment by binji** [comment]
@nta Agreed that the spec can be difficult to read at times. We have added explanatory notes in places, but perhaps there could be more. And I'm sure there are places we can clarify the prose as well. That said, the spec is unlikely to be changed to pseudo-code, for the reasons others have mentioned above. I know that some have found this document clearer: https://github.com/sunfishcode/wasm-reference-manual. It hasn't been updated in a little while, but I believe the information there is still up-to-date with the WebAssembly MVP.

**Comment by Choons** [comment]
in 100% agreement with original poster. I got an 'A' in college calculus, but that spec went over my head about 30 pages in. Like trying to read a math book. I understand the need for specificity, but that stuff should be in the Appendix for reference.

**Comment by fwsGonzo** [comment]
Why does the spec look like this?

**Comment by ngzhian** [comment]
@fwsGonzo Are there particular examples in that PDF that you would like to point to as a good example of how a manual should be written?

> Why does the spec look like this?

Which part of the spec were you reading when you had this thought? Maybe it was a poorly written section that we can polish. Is your comment related to formal text, math equations, or even the styling? We want to make the specification more appetizing, so more specific feedback is welcome.

If this is more of your overall feeling, refer to this comment by @gahaas above:

> The spec is written both in prose as well as with the mathematical notations. There is no need to read the mathematical notations, just stick to the prose text. When you implement the spec, then the prose text is sometimes unclear. In that case, the mathematical notations bring clarity.

I agree with his sentiment - I have been working on adding spec text to the WebAssembly/simd proposal, and the formal mathematics have helped me spot some bugs in our semantics. The more concise mathematical notation sometimes make it easier for me to pattern-match and notice when things are out of place. When looking at other sections, my eyes gloss over the math, and I'll read the prose to find out what I need to.

**Comment by fwsGonzo** [comment]
I don't doubt you at all. It's just a completely different way to write a spec that an "old" fart doesn't understand at all. :) 
I did find an alternative spec in another repo that resembles more what I'm used to, so I'll probably be fine.

More than anything, prose is less important than explaining the bits and bytes and often repeating yourself. Specs that try to DRY things out by referring back to things could instead just auto-inline these references instead. I say that, because I'm used to reading specs that are one-and-done, and because of that they don't mind repeating some fixed value, because it never changes. I suppose that is not necessarily true here. I don't know. Either way, I will try to figure out the spec.

Have a nice day.
